mana_decay_provider_test.go

```diff
@@ -368,3 +390,62 @@ func TestManaDecay_Rewards(t *testing.T) {
 		})
 	}
 }
+
+// TestFloatManaDecayProvider calculates the mana decay and mana generation
+// using floating point arithmetic.
+type TestFloatManaDecayProvider struct {
+	timeProvider *iotago.TimeProvider
+
+	// generationRate is the amount of potential Mana generated by 1 IOTA in 1 slot.
+	generationRate uint64 // the generation rate needs to be scaled by 2^-generationRateExponent
+
+	// generationRateExponent is the scaling of generationRate expressed as an exponent of 2.
+	generationRateExponent uint64
+}
+
+func (p *TestFloatManaDecayProvider) ManaGenerationWithDecay(amount iotago.BaseToken, creationSlot iotago.SlotIndex, targetSlot iotago.SlotIndex) float64 {
+	creationEpoch := p.timeProvider.EpochFromSlot(creationSlot)
+	targetEpoch := p.timeProvider.EpochFromSlot(targetSlot)
+	floatAmount := float64(amount)
+	floatGenerationRate := float64(p.generationRate) * math.Pow(2, -float64(p.generationRateExponent))
+	delta := float64(1<<13) * (1.0 / (365.0 * 24.0 * 60.0 * 60.0)) * float64(10)
+	epochDiff := targetEpoch - creationEpoch
+
+	//nolint:exhaustive // false-positive, we have a default case
+	switch epochDiff {
+	case 0:
+		floatSlotDiff := float64(targetSlot - creationSlot)
+
+		return floatSlotDiff * floatAmount * floatGenerationRate
+
+	case 1:
+		slotsBeforeNextEpoch := p.timeProvider.SlotsBeforeNextEpoch(creationSlot)
+		slotsSinceEpochStart := p.timeProvider.SlotsSinceEpochStart(targetSlot)
+		manaDecayed := float64(slotsBeforeNextEpoch) * floatAmount * floatGenerationRate * math.Exp(-delta/3)
+		manaGenerated := float64(slotsSinceEpochStart) * floatAmount * floatGenerationRate
+
+		return manaDecayed + manaGenerated
+
+	default:
+		slotsBeforeNextEpoch := p.timeProvider.SlotsBeforeNextEpoch(creationSlot)
+		slotsSinceEpochStart := p.timeProvider.SlotsSinceEpochStart(targetSlot)
+		epochDiffFloat := float64(epochDiff)
+		slotsPerEpochFloat := math.Pow(2, 13)
+		constant := math.Exp(-delta/3) * (1 - math.Exp(-(epochDiffFloat-1)*delta/3)) / (1 - math.Exp(-delta/3))
+		potentialMana_n := float64(slotsBeforeNextEpoch) * floatAmount * floatGenerationRate * math.Exp(-epochDiffFloat*delta/3)
+		potentialMana_n_1 := constant * floatAmount * floatGenerationRate * slotsPerEpochFloat
+		potentialMana_0 := float64(slotsSinceEpochStart) * floatAmount * floatGenerationRate
+
+		return potentialMana_n + potentialMana_n_1 + potentialMana_0
+	}
+}
+
+func (p *TestFloatManaDecayProvider) LowerBoundPotentialMana(amount iotago.BaseToken, creationSlot iotago.SlotIndex, targetSlot iotago.SlotIndex) float64 {
```

Comment:
You added a review comment yourself before I moved that code:
`explain where these bounds came from`
Is this still a thing?

Reply:
let's leave it like it is